reference/zrc6.scilla

```diff
@@ -0,0 +1,908 @@
+    _recipient: _sender;
+    _amount: Uint128 0;
+    to: to
+  };
+  msgs = one_msg msg_to_sender;
+  send msgs
+end
+
+(* Adds `to` as spender of `token_id`. *)
+(* There can only be one spender per token at any given time. *)
+(* @Requirements: *)
+(* - `token_id` should exist. Otherwise, it should throw `TokenNotFoundError` *)
+(* - `_sender` should be a token owner or an operator. Otherwise, it should throw `NotOwnerOrOperatorError` *)
+(* - `to` should not be already a spender. Otherwise, it should throw `SpenderFoundError` *)
+transition AddSpender(to: ByStr20, token_id: Uint256)
+  opt_token_owner <- token_owners[token_id];
+  match opt_token_owner with
+  | None =>
+    error = TokenNotFoundError;
+    Throw error
+  | Some token_owner =>
+    RequireOwnerOrOperator token_owner;
+    
+    has_spender <- exists spenders[token_id];
+    match has_spender with
+      | True =>
+      error = SpenderFoundError;
+      Throw error
+      | False =>
+      (* Add spender *)
+      spenders[token_id] := to
+    end;
+    
+    e = {
+      _eventname: "AddSpender";
+      to: to;
+      token_id: token_id
+    };
+    event e;
+    msg_to_sender = { 
+      _tag: "ZRC6_AddSpenderCallback";
+      _recipient: _sender;
+      _amount: Uint128 0;
+      to: to;
+      token_id: token_id
+    };
+    msgs = one_msg msg_to_sender;
+    send msgs
+  end
+end
+
+(* Removes `to` from spender of `token_id`. *)
+(* @Requirements: *)
+(* - `token_id` should exist. Otherwise, it should throw `TokenNotFoundError` *)
+(* - `_sender` should be a token owner or an operator. Otherwise, it should throw `NotOwnerOrOperatorError` *)
+(* - `to` should be already a spender. Otherwise, it should throw `SpenderNotFoundError` *)
+transition RemoveSpender(to: ByStr20, token_id: Uint256)
+  opt_token_owner <- token_owners[token_id];
+  match opt_token_owner with
+  | None =>
+    error = TokenNotFoundError;
+    Throw error
+  | Some token_owner =>
+    RequireOwnerOrOperator token_owner;
+    
+    has_spender <- exists spenders[token_id];
+    match has_spender with
+    | True =>
+      (* Remove spender *)
+      delete spenders[token_id]
+    | False =>
+      error = SpenderNotFoundError;
+      Throw error
+    end;
+    
+    e = {
+      _eventname: "RemoveSpender";
+      to: to;
+      token_id: token_id
+    };
+    event e;
+    msg_to_sender = { 
+      _tag: "ZRC6_RemoveSpenderCallback";
+      _recipient: _sender;
+      _amount: Uint128 0;
+      to: to;
+      token_id: token_id
+    };
+    msgs = one_msg msg_to_sender;
+    send msgs
+  end
+end
+
+(* Adds `to` as operator for the `_sender`. *)
+(* @Requirements: *)
+(* - `_sender` should be the token owner. Otherwise, it should throw `NotTokenOwnerError` *)
+(* - `_sender` should not be `to`. Otherwise, it should throw `SelfError` *)
+(* - `to` should not be already an operator. Otherwise, it should throw `OperatorFoundError` *)
+transition AddOperator(to: ByStr20)
```

Comment:
can consider renaming `to` to `address`

Reply:
Let's go with `operator` as we discussed.